besspin/cyberPhys/cyberphyslib/cyberphyslib/demonstrator/health.py

```diff
@@ -0,0 +1,334 @@
+    """
+    def __init__(self, addr: str, user=None, password=None, **kwargs):
+        # user and password need to be transformed into proper level of fabric
+        # kwarg hierarchy
+        if user:
+            kwargs["user"] = user
+        if password:
+            if "connect_kwargs" in kwargs:
+                kwargs["connect_kwargs"] = {**{"password": password}, **kwargs}
+            else:
+                kwargs["connect_kwargs"] = {"password": password}
+        self.connection_kwargs = kwargs
+        self.addr: str = addr
+
+    def command(self, command: str):
+        # NOTE: should I persist the connect rather than make a new one each time?
+        result = fabric.Connection(self.addr, **self.connection_kwargs).run(command, hide=True)
+        return result
+
+
+class HttpMonitor(HealthMonitor):
+    """
+    Health monitoring involving http connections
+    """
+    def __init__(self, addr: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.addr = addr
+
+    @abc.abstractmethod
+    def check_response(self, ret: requests.Response) -> bool:
+        raise NotImplementedError
+
+    def run_health_test(self) -> bool:
+        try:
+            ret = requests.get(self.addr)
+            # check for OK status code
+            return self.check_response(ret)
+        except Exception as exc:
+            # TODO: log this?
+            return False
+
+
+class ServiceMonitor(SshMonitor):
+    """
+    Health monitoring of a service on a remote machine (via Ssh)
+    """
+    def __init__(self, service_name: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.service_name = service_name
+
+    def run_health_test(self) -> bool:
+        try:
+            ret = self.command(rf"systemctl status {self.service_name}")
+            # regex match for active and running service
+            # NOTE: is this correct and stable?
+            return re.search(r"(Active: active \(running\))", ret.stdout) is not None
+        except Exception as exc:
+            # TODO: log this?
+            print(f"Exception {exc} has occurred")
+            return False
+
+
+class OtaMonitor(HttpMonitor):
+    """
+    Monitor Infotainment Ota Process (via HTTP request)
+    """
+    def check_response(self, ret: requests.Response) -> bool:
+        return ret.status_code == 200
+
+
+HeartbeatResponseType = typing.Union[None, typing.Tuple[int, int]]
+
+
+class TcpHeartbeatMonitor(HealthMonitor):
+    """
+    Heartbeat monitor is responsible for
+
+        1. sending out the heartbeat requests (REQ)
+        2. listening for heartbeat acknowledgements (ACK)
+        3. determining system health from ACK patterns
+    """
+    # length of response buffers (time horizon of ACKs)
+    maxlen = 5
+
+    @staticmethod
+    def form_heartbeat_req_msg(req_number: int):
+        """form REQ can packet"""
+        msg = can.Message(arbitration_id=canspecs.CAN_ID_HEARTBEAT_REQ,
```

Comment:
not sure how important is to explicitly set `DLC` - can Message figure it out based on the `data`?